### ISSUE TITLE
Interface for providing fixes for diagnostics

### ISSUE THREAD

**Issue by rictic** [Microsoft/language-server-protocol#144]
Some diagnostics indicate clear errors with unambiguous and well defined fixes. e.g. a missing semicolon in javascript.

As I read the spec today, the flow for providing a fix for a diagnostic is more or less:

1. Server publishes diagnostics
2. Tool requests code action based on user interaction with one or more diagnostics
3. Server responds with `Command[]`
4. Tool does something implementation-specific with commands

Language servers are doing ad hoc work beyond this. The tslint vscode plugin for example has a private extension to the protocol to gather all fixes for a file as a `TextEdit[]`, and can be configured to apply all fixes on save. 

It would be valuable for both server and tool authors to have a specification for this class of use cases.

**Comment by rictic** [comment]
A few thoughts on what the interface might look like:

We could represent a fix as either a WorkspaceEdit a TextEdit[]. I think there's a case to be made for a WorkspaceEdit, as there are some reasonable fixes aren't local to a single file. For example a linter that applies a naming convention to fields could have a fix that involves renaming the field, requiring changes across multiple files.

We could add `fix` as an optional property on Diagnostic. It seems worthwhile to add an optional resolve step analogous to the one for completion items, in cases where computing the fix may be expensive. 

This does make conflict resolution less clear though. If a tool wants to apply all fixes, but there are overlaps in the TextEdits (between diagnostics, not in the TextEdits of a single diagnostic, as that is obviously invalid) what should it do? It could go fix by fix, building up a larger fix by only accepting fixes that do not overlap with any of the TextEdits accepted so far.

Alternatively, we could add a new request for a fix and let the server handle conflict resolution. This interface ends up looking a lot like the existing format document request though.

**Comment by dbaeumer** [comment]
Here is what I added for 3.0 to support this:

- clients can announce a capability that they support applyEdit requests
- sever can announce a capability that they support command execution.

Using the two new requests auto fix is implemented like this:

- server publishes diagnostics
- client request code actions / commands for the diagnostics. Server provides them
- User picks a command
- clients forwards the execution to the server
- server computes edits and send a applyEdit request to the client

The eslint extension client part has no custom requests anymore to support auto fix.

@rictic would that work for you as well.

**Comment by dbaeumer** [comment]
And to note: the edit that can be applied using the new applyEdit request support modifying n files.

**Comment by dbaeumer** [comment]
OK to close ?

**Comment by rictic** [comment]
Yeah, looks good to me. So, to summarize, the manual flow works as you described, and for an automatic fix flow the server can choose to either respond to `textDocument/willSaveWaitUntil` with TextEdits or send down `workspace/applyEdit`? 

That should work great, I'll try it out soon!

**Comment by rictic** [comment]
Ok, so, it's only nine months later, but I got this working!

The flow is surprisingly complicated and mind-twisty. I think it would be worth adding some more documentation to the suggested flow for fixing lint warnings. For the code action part I eventually grokked what you were saying that the flow is:
```
  <- publishDiagnostics
  -> getCodeActions  <- Command(myAppName/applyEdit)
  -> executeCommand(myAppName/applyEdit)
  <- workspace/applyEdit
```

It is very nice though that I can execute other commands without any code needed in the client-side part of the extension. +1 to that.

The only remaining issue here that I've got is Microsoft/vscode-languageserver-node#274
